keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2868 @@
+matching binding. In the above example, the "nontest" policy binding could
+instead have been defined as a global policy:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyBinding
+metadata:
+  name: "replicalimit-binding-global"
+spec:
+  policy: "replicalimit-policy.example.com"
+  params: "replica-limit-clusterwide.example.com"
+  matchResources:
+    namespaceSelectors:
+    - key: environment,
+      operator: Exists
+  enforcement: [Deny]
+```
+
+With this binding, the test and global policy bindings overlap. Resources
+admitted to test environment would then be checked against both policy
+configurations.
+
+To summarize, this "API Shape" separates admission policy _definition_ from
+_configuration_. This has a couple advantages:
+
+- Access to, and delegation of, policy configuration is more manageable. In
+  particular, Kubernetes RBAC works well with this design.
+- Without the separation, the next most obvious API shape would be to encode
+  everything into a single resource, which could easily become very large and
+  run into resource size limits.
+
+##### Policy Definitions
+
+Policy definitions are responsible for:
+
+- Defining what validations the policy enforces and how violations are reported 
+- Defining how a policy may be configured
+
+Each `ValidatingAdmissionPolicy` resource contains a `spec.matchConstraints` to
+declare what resources it validates. This field is required.
+
+- `spec.matchConstraints` constrains which resources this policy can be applied
+  to. Policy bindings each have match rules with further narrow this constraint,
+  but cannot expand it. This allows the CEL expressions to make safe
+  assumptions. E.g. a CEL expression that is constrained to CREATES and UPDATES
+  of resources is guaranteed the root `object` variable is never null, but a CEL
+  expression that might need to evaluate a DELETE must handle the root `object`
+  variable being null. See below "Match criteria" section for how match criteria
+  is described.
+- `spec.matchConstraints` guides CEL expression type checking, see the below
+  "Type safety" section for more details. 
+
+CEL expressions have access to the contents of the `AdmissionReview` type,
+organized into CEL variables as well as some other useful variables:
+
+- 'object'
+- 'oldObject'
+- 'review'
+  - 'requestResource' (GVR)
+  - 'resource' (GVR)
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated
+
+See below "Reporting violations to Clients" for more detail about how the
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configuration
+
+`PolicyBinding` resources and parameter CRDs together define how cluster
+administrators configure policies for clusters.
+
+Each `PolicyBinding` contains:
+
+- `spec.policy` - A reference to the policy being configured
```

Comment:
Unless I'm misunderstanding, this can be supported via:

- policy A <- binding X -> params 1
- policy B <- binding Y -> params 1

Reply:
This is true, my point is that it would require multiple bindings. I.e. adding a policy def that extends an existing policy doesn't get it enforced for "free".

If needed, maybe that could be added later by not specifying `spec.policy` or similar, but worth highlighting the tradeoff.

Reply:
I think the params object can't reference the policy object, since it could work with multiple.

The policy object has to reference the params *type* (but not a specific param instance) in order to type-check.

So yeah, it seems like you have to state both in order to have a well defined binding.

Default behavior of e.g. "blank means all param objects" is not that user friendly, if we absolutely need it I would make it require a "*" or something.

Reply:
Although... without a params object, you also can't have multiple policy objects reference the same config (since this is done via listening to a common kind).

Does this imply creating empty params objects for supporting this functionality?

Or does it mean that bindings must double-cite and we lose the "multiple policies read the same config just by existing" behavior?

Or does bindings-cite-policies become optional, which makes for a bit more heterogeneity in terms of the possible config footprints?